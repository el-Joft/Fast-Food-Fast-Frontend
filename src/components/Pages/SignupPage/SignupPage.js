/* eslint-disable react/no-unused-state */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import './SignupPage.scss';
import { registerUser } from '../../../actions/authActions';

class SignupPage extends Component {
  state = {
    firstname: '',
    lastname: '',
    email: '',
    phone: '',
    city: '',
    zipcode: '',
    address: '',
    password: '',
    // confirmPassword: '',
    errors: {},
  }

  // this prevents the user from accessing the login page when authenticated
  componentDidMount() {
    if (this.props.auth.isAuthenticated) {
      this.props.history.push('/');
    }
  }

  // here we are checking if the component recieves a props
  // if it does then the props must contain the errors
  // if it does the it should set the state of errors to nextprops.errors
  componentWillReceiveProps(nextProps) {
    if (nextProps.auth.isAuthenticated) {
      nextProps.history.push('/');
    }

    if (nextProps.errors) {
      this.setState({
        errors: nextProps.errors,
      });
    }
  }

  onChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  onSubmit = (event) => {
    event.preventDefault();
    const newUser = {
      firstname: this.state.firstname,
      lastname: this.state.lastname,
      email: this.state.email,
      phone: this.state.phone,
      city: this.state.city,
      zipcode: this.state.zipcode,
      address: this.state.address,
      password: this.state.password,
      // confirmPassword: this.state.confirmPassword,
    };
    this.props.registerUser(newUser);
  };

  render() {
    const { user } = this.props.auth;
    return (
      <section>
        {user ? user.firstname : '' }
        { this.state.errors === ''
          ? <div className='message' />
          : <div className='error_message'>{this.state.errors.message}</div>
      }
        <div>
          <form onSubmit={this.onSubmit}>
            <ul className='form-group'>
              <li>
                <label>
Full Name
                  {' '}
                  <span className="required">*</span>
                </label>
                <input onChange={this.onChange} value={this.state.firstname} type="text" name="firstname" className="form-control field-divided" placeholder="First Name" />
&nbsp;
                <input onChange={this.onChange} value={this.state.lastname} type="text" name="lastname" className="form-control field-divided" placeholder="Last Name" />
              </li>
              <li>
                <label>
Email
                  {' '}
                  <span className="required">*</span>
                </label>
                <input onChange={this.onChange} value={this.state.email} type="email" name="email" className="form-control field-long" required />
              </li>
              <li>
                <label>
Phone
                  {' '}
                  <span className="required">*</span>
                </label>
                <input onChange={this.onChange} value={this.state.tphone} type="text" name="phone" className="form-control field-long" required />
              </li>
              <li>
                <label>
City
                  {' '}
                  <span className="required">*</span>
                </label>
                <input onChange={this.onChange} value={this.state.city} type="text" name="city" className="form-control field-long" required />
              </li>
              <li>
                <label>
Zip Code
                  {' '}
                  <span className="required">*</span>
                </label>
                <input onChange={this.onChange} value={this.state.zipcode} type="number" name="zipcode" className="form-control field-long" required />
              </li>
              <li>
                <label>
Address
                  {' '}
                  <span className="required">*</span>
                </label>
                <textarea onChange={this.onChange} value={this.state.address} name="address" id="address" rows="3" required="required" className="field-long field-textarea" />
              </li>
              <li>
                <label>
Password
                  {' '}
                  <span className="required">*</span>
                </label>
                <input onChange={this.onChange} value={this.state.password} type="password" name="password" className="form-control field-long" required />
              </li>
              {/* <li>
                <label>
Confirm Password
                  {' '}
                  <span className="required">*</span>
                </label>
                <input onChange={this.onChange} value={this.state.confirmPassword} type="password" name="confirmPassword" className="form-control field-long" required />
              </li> */}

              <li>
                <input type="submit" value="Submit" />
              </li>
            </ul>
          </form>
          <div className="text-form">
            <p>
              Already have an account?
              {' '}
              <a href="signin.html">Login</a>
            </p>
          </div>
        </div>

      </section>
    );
  }
}

SignupPage.propTypes = {
  registerUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired,
};

// we get values here using the this.props.errors/auth

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors,
});

export default connect(mapStateToProps, { registerUser })(withRouter(SignupPage));
