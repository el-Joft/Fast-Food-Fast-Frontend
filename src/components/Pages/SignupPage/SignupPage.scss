// /* FORM */

// .form-control {
//   width: 30%;
//   padding: 10px;
//   margin-bottom: 10px;
//   font-size: 1rem;
//   line-height: 1.5;
//   color: #495057;
//   background-color: #fff;
//   border: 1px solid #42ce9f;
//   border-radius: .25rem;
//   transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
// }

// .form-group {
//   margin:10px auto;
//   max-width: 400px;
//   padding: 20px 12px 10px 20px;
//   font: 13px "Lucida Sans Unicode", "Lucida Grande", sans-serif;
// }
// .form-group li {
//   padding: 0;
//   display: block;
//   list-style: none;
//   margin: 10px 0 0 0;
// }
// .form-group label{
//   margin:0 0 3px 0;
//   padding:0px;
//   display:block;
//   font-weight: bold;
//   font-size: 1.7;
// }

// .select {
//   border: 1px solid #42ce9f;
// }
// option {
//   color: #42ce9f;
// }

// .form-group .field-long{
//   width: 100%;
// }

// .form-group input[type=submit], .form-group input[type=button]{
//   background: #42ce9f;
//   padding: 8px 15px 8px 15px;
//   border: none;
//   color: #fff;
// }
// .form-group .field-divided{
//   width: 49%;
// }
// .form-group .required{
//   color:red;
// }
// /* this aspect was learnt from the MDN */
// /* https://developer.mozilla.org/en-US/docs/Learn/HTML/Forms/Styling_HTML_forms */
// /* MORE ON FORMS */
// .form-group input[type=text]:focus, 
// .form-group input[type=date]:focus,
// .form-group input[type=datetime]:focus,
// .form-group input[type=number]:focus,
// .form-group input[type=search]:focus,
// .form-group input[type=time]:focus,
// .form-group input[type=url]:focus,
// .form-group input[type=email]:focus,
// .form-group input[type=password]:focus,
// .form-group textarea:focus, 
// .form-group select:focus{
//   -moz-box-shadow: 0 0 8px #88D5E9;
//   -webkit-box-shadow: 0 0 8px #88D5E9;
//   box-shadow: 0 0 8px #88D5E9;
//   border: 1px solid #88D5E9;
// }

// /* end of the aspect */



// .text-form{
//   text-align: center;
//   padding: 10px;

// }
