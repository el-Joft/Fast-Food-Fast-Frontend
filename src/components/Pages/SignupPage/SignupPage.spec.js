/* eslint-disable no-undef */
/* eslint-disable react/react-in-jsx-scope */
import { shallow } from 'enzyme';
import { SignupPage } from './SignupPage';

describe('rendering', () => {
  let wrapper;
  let props;
  let nextProps;
  beforeEach(() => {
    props = {
      registerUser: jest.fn(),
      history: {
        push: jest.fn(),
      },
      success: {
        status: true,
      },
      orders: [],
      auth: {
        isAuthenticated: true,
        user: {
          id: 1,
        },
      },
    };
    nextProps = {
      registerUser: jest.fn(),
      history: {
        push: jest.fn(),
      },
      success: {
        status: true,
      },
      orders: [],
      auth: {
        isAuthenticated: true,
        user: {
          id: 1,
        },
      },
    };
  });

  it('invokes `componentDidMount` when mounted', () => {
    const spy = jest.spyOn(SignupPage.prototype, 'componentDidMount');
    wrapper = shallow(<SignupPage {...props} />);
    wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  it('invokes `componentWillRecieveProps` when mounted', () => {
    const spy = jest.spyOn(SignupPage.prototype, 'componentWillReceiveProps');
    wrapper = shallow(<SignupPage {...props} />);
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(spy).toHaveBeenCalled();
  });

  it('should render the component', () => {
    wrapper = shallow(<SignupPage {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the Button component', () => {
    wrapper = shallow(<SignupPage {...props} />);
    expect(wrapper.length).toBe(1);
  });
});
