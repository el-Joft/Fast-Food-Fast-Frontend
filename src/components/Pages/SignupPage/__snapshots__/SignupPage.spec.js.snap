// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render the component 1`] = `
<section>
  <div
    className="error_message"
  />
  <div>
    <form
      onSubmit={[Function]}
    >
      <ul
        className="form-group"
      >
        <li>
          <label>
            Full Name
             
            <span
              className="required"
            >
              *
            </span>
          </label>
          <input
            className="form-control field-divided"
            name="firstname"
            onChange={[Function]}
            placeholder="First Name"
            type="text"
            value=""
          />
          Â 
          <input
            className="form-control field-divided"
            name="lastname"
            onChange={[Function]}
            placeholder="Last Name"
            type="text"
            value=""
          />
        </li>
        <li>
          <label>
            Email
             
            <span
              className="required"
            >
              *
            </span>
          </label>
          <input
            className="form-control field-long"
            name="email"
            onChange={[Function]}
            required={true}
            type="email"
            value=""
          />
        </li>
        <li>
          <label>
            Phone
             
            <span
              className="required"
            >
              *
            </span>
          </label>
          <input
            className="form-control field-long"
            name="phone"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </li>
        <li>
          <label>
            City
             
            <span
              className="required"
            >
              *
            </span>
          </label>
          <input
            className="form-control field-long"
            name="city"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </li>
        <li>
          <label>
            Zip Code
             
            <span
              className="required"
            >
              *
            </span>
          </label>
          <input
            className="form-control field-long"
            name="zipcode"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </li>
        <li>
          <label>
            Address
             
            <span
              className="required"
            >
              *
            </span>
          </label>
          <textarea
            className="field-long field-textarea"
            id="address"
            name="address"
            onChange={[Function]}
            required="required"
            rows="3"
            value=""
          />
        </li>
        <li>
          <label>
            Password
             
            <span
              className="required"
            >
              *
            </span>
          </label>
          <input
            className="form-control field-long"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
        </li>
        <li>
          <input
            type="submit"
            value="Submit"
          />
        </li>
      </ul>
    </form>
    <div
      className="text-form"
    >
      <p>
        Already have an account?
         
        <a
          href="signin.html"
        >
          Login
        </a>
      </p>
    </div>
  </div>
</section>
`;
